File: src/test/scala/com/github/mauricio/async/db/postgresql/DatabaseTestHelper.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.postgresql

import com.github.mauricio.async.db.Configuration
import concurrent.Await
import concurrent.duration._

trait DatabaseTestHelper {

  def databaseName = Some("netty_driver_test")

  def databasePort = 5433

  def defaultConfiguration = new Configuration(
    port = databasePort,
    username = "postgres",
    database = databaseName)

  def withHandler[T](fn: (DatabaseConnectionHandler) => T): T = {
    withHandler(this.defaultConfiguration, fn)
  }

  def withHandler[T](configuration: Configuration, fn: (DatabaseConnectionHandler) => T): T = {

    val handler = new DatabaseConnectionHandler(configuration)

    try {
      Await.result(handler.connect, Duration(5, SECONDS))
      fn(handler)
    } finally {
      handler.disconnect
    }

  }

  def executeDdl(handler: DatabaseConnectionHandler, data: String, count: Int = 0) = {
    val rows = Await.result(handler.sendQuery(data), Duration(5, SECONDS)).rowsAffected

    if (rows != count) {
      throw new IllegalStateException("We expected %s rows but there were %s".format(count, rows))
    }

  }

  def executeQuery(handler: DatabaseConnectionHandler, data: String) = {
    Await.result(handler.sendQuery(data), Duration(5, SECONDS))
  }

  def executePreparedStatement(
                                handler: DatabaseConnectionHandler,
                                statement: String,
                                values: Array[Any] = Array.empty[Any]) = {
    Await.result(handler.sendPreparedStatement(statement, values), Duration(5, SECONDS))
  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/DatabaseTestHelper.kt
package com.github.aysnc.sql.db.integration

import com.github.jasync.sql.db.Configuration
import com.github.jasync.sql.db.Connection
import com.github.jasync.sql.db.QueryResult
import com.github.jasync.sql.db.SSLConfiguration
import com.github.jasync.sql.db.postgresql.PostgreSQLConnection
import io.netty.handler.timeout.TimeoutException
import java.io.File
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

open class DatabaseTestHelper {

    private val containerHelper = ContainerHelper

    val conf = containerHelper.defaultConfiguration

    fun <T> withHandler(fn: (PostgreSQLConnection) -> T): T {
        return withHandler(containerHelper.defaultConfiguration, fn)
    }

    val defaultSslConfig = SSLConfiguration(
        mode = SSLConfiguration.Mode.Require,
        rootCert = File(ClassLoader.getSystemClassLoader().getResource("server.cert.txt").file)
    )

    fun <T> withSSLHandler(
        host: String = "localhost",
        sslConfig: SSLConfiguration = defaultSslConfig,
        fn: (PostgreSQLConnection) -> T
    ): T {
        val config = containerHelper.defaultConfiguration.copy(
            host = host,
            ssl = sslConfig
        )
        return withHandler(config, fn)
    }

    fun <T> withHandler(configuration: Configuration, fn: (PostgreSQLConnection) -> T): T {
        val handler = PostgreSQLConnection(configuration)

        try {
            handler.connect().get(5, TimeUnit.SECONDS)
            return fn(handler)
        } finally {
            handleTimeout(handler) { handler.disconnect() }
        }
    }

    fun executeDdl(handler: Connection, data: String, count: Int = 0): Long {
        val rows = handleTimeout(handler) {
            handler.sendQuery(data).get(5, TimeUnit.SECONDS).rowsAffected
        }

        if (rows.toInt() != count) {
            throw IllegalStateException("We expected %s rows but there were %s".format(count, rows))
        }

        return rows
    }

    private fun <R> handleTimeout(handler: Connection, fn: () -> R): R {
        try {
            return fn()
        } catch (e: TimeoutException) {
            throw IllegalStateException("Timeout executing call from handler -> %s".format(handler))
        }
    }

    fun executeQuery(handler: Connection, data: String): QueryResult {
        return handleTimeout(handler) {
            handler.sendQuery(data).get(5, TimeUnit.SECONDS)
        }
    }

    fun executePreparedStatement(
        handler: Connection,
        statement: String,
        values: List<Any?> = emptyList(),
        release: Boolean = false
    ): QueryResult {
        return handleTimeout(handler) {
            handler.sendPreparedStatement(statement, values, release).get(5, TimeUnit.SECONDS)
        }
    }

    fun <T> awaitFuture(future: CompletableFuture<T>): T {
        return future.get(5, TimeUnit.SECONDS)
    }

    fun releasePreparedStatement(handler: PostgreSQLConnection, query: String) {
        return handleTimeout(handler) {
            awaitFuture(handler.releasePreparedStatement(query))
        }
    }
}
